### PR TITLE
Enable merge option for datasets (phase 1)

### PR DESCRIPTION
**Pull request by Sigmonia**
#### Rationale
Spec: https://docs.google.com/document/d/110fl8qRIk1o9lEgFq8RKFWNW5vo_cYCiWkE1doFq214/
Enable the Merge option for bulk updating Datasets

TODO: (Phase 2) Improve Audit Logging

#### Changes
* Enabled UI option for the Dataset bulk insert page and form
* Passed importOption and auditLogBehavior parameters through to the DataIterator context
* Added a DetailedAuditLogDataIterator to the DI queue

### CODE REVIEW

**Comment by Sigmonia** [comment]
One thing I just noticed is the Created date field's value is getting changed on the update instead of the Modified field. I'll need to look and see how to correct it.

File: study/src/org/labkey/study/model/DatasetDefinition.java [hunk 1]
```diff
@@ -1586,6 +1590,35 @@ public PHI getMaxContainedPhi()
...
+            }
+        }
+
+        @Override
+        public void addAuditEvent(User user, Container container, AuditBehaviorType auditBehavior, @Nullable String userComment, QueryService.AuditAction auditAction, Map<String, Object> parameters)
+        {
+            Map<String,Object> oldRow = null;
+            if (auditAction == QueryService.AuditAction.MERGE)
+            {
+                //REVIEW: I'm guessing this is a terrible idea for multiple reasons, but seems to work...
```
Comment (labkey-susanh) [inline study/src/org/labkey/study/model/DatasetDefinition.java:1615]: This is certainly where we take the performance hit, but it is akin to how we handle the detailed audit logging for samples as well and is, I believe, the best we can do without more significant changes.  What reasons were you seeing other than performance for this to be a bad idea?

File: study/src/org/labkey/study/model/DatasetDefinition.java [hunk 2]
```diff
@@ -2096,20 +2129,26 @@ public DataIteratorBuilder getInsertDataIterator(User user, DataIteratorBuilder
         TableInfo table = getTableInfo(user, false);
         boolean needToHandleQCState = table.getColumn(DatasetTableImpl.QCSTATE_ID_COLNAME) != null;
 
+        DataIteratorBuilder dib = in;
```
Comment (labkey-susanh) [inline study/src/org/labkey/study/model/DatasetDefinition.java:2132]: The extra variable here doesn't seem necessary.

File: study/src/org/labkey/study/model/DatasetDefinition.java [hunk 3]
```diff
@@ -1586,6 +1590,35 @@ public PHI getMaxContainedPhi()
...
+
+                StudyServiceImpl.addDatasetAuditEvent(user, DatasetDefinition.this, row, updatedRow);
+            }
+        }
+
+        @Override
+        public void addAuditEvent(User user, Container container, AuditBehaviorType auditBehavior, @Nullable String userComment, QueryService.AuditAction auditAction, Map<String, Object> parameters)
+        {
+            Map<String,Object> oldRow = null;
+            if (auditAction == QueryService.AuditAction.MERGE)
```
Comment (labkey-susanh) [inline study/src/org/labkey/study/model/DatasetDefinition.java:1613]: We want to do the detailed logging only when the audit behavior is DETAILED.

**Review by labkey-susanh — CHANGES_REQUESTED** [review]

**Comment by Sigmonia** [comment]
Split into two PRs:
1) for enabling the merge option https://github.com/LabKey/platform/pull/1839
2) Adding the audit logging https://github.com/LabKey/platform/pull/1840

**Pull request closed without merging**